keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+    // which attributes.
+    Status ResourceClaimStatus
+}
+
+type ResourceClaimSpec struct {
+    // ResourceClassName references the driver and additional
+    // parameters via the name of a ResourceClass that was
+    // created as part of the driver deployment.
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a driver-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
+
+    // Allocation can start immediately or when a Pod wants to use
+    // the resource. Waiting for a Pod is the default.
+    AllocationMode AllocationMode
+}
+
+type AllocationMode string
+
+const (
+     AllocationModeImmediate AllocationMode = “Immediate”
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the node candidates triggers a check
+   // on which nodes the resource could be made available.
+   // Unsuitable nodes will be ignored when selecting
+   // a node. All other nodes are potential candidates,
+   // either because no information is available yet
+   // or because allocation is expected to succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically until a node gets
+   // selected by the scheduler.
+   UnsuitableNodes []string
+
+   // An allocated resource is available on nodes that match this
+   // selector. If nil, the resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
```

Comment:
It only has a meaning during "phase=allocated". The resource driver can unset it after deallocation and probably should to keep the object smaller, but it won't be required. I'll clarify.

Reply:
Actually, I think we can relax that and make this field apply also to ResourceClaims in the Pending phase. kube-scheduler then will use that field, if non-nil, during Filter to narrow down the choice, in addition to checking UnsuitableNodes.

This may be useful for resource drivers that know in advance that their resources will only be available on nodes with certain labels. Later, when setting phase=allocated, they can narrow it down to the actual node(s) for which the resource was allocated.

Reply:
Updated accordingly.

Reply:
Is the driver expected to add/remove labels from nodes?

Reply:
No. It may decide to do that, but that's really an implementation detail.

For example, a specific resource driver might instead rely on Node Feature Discovery (NFD) to do the labeling and then only run on suitable nodes. Then this node selector can also use those labels.

For resources that are only available on a single node the node name can be used instead of labels.